(#12430) Improve detection of mutable value types

vexx32 opened the pull request:
# PR Summary

Currently, we have some flawed assumptions around what defines a mutable value type. Rather than make a more complicated series of checks and making assumptions, we can simply look at whether the type is defined with the `IsReadOnlyAttribute` that the compiler uses to guarantee a value type is immutable.

As a fallback, it's much safer and less complex to enumerate all public and private fields on the type, and verify whether they're all read-only.

## PR Context

Resolves #12430. See the issue for discussion and further context.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

PoshChan commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -2029,36 +2029,34 @@ private static object NullRule(CallSite site, object obj)
 
```
@vexx32, your last commit had 1 failures in `PowerShell-CI-linux`
TabCompletion.Miscellaneous completion tests.Test history completion
```powershell
Expected strings to be the same, but they were different.
Expected length: 23
Actual length:   478
Strings differ at index 0.
Expected: 'Test history completion'
But was:  '$env:POWER...'
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1: line 909
909:             $res.CompletionMatches[0].CompletionText | Should -BeExactly "Test history completion"
```

PoshChan replied:
@vexx32, your last commit had 3 failures in `PowerShell-CI-macos`
TabCompletion.Miscellaneous completion tests.Test history completion
```powershell
Expected strings to be the same, but they were different.
Expected length: 23
Actual length:   598
Strings differ at index 0.
Expected: 'Test history completion'
But was:  '$env:POWER...'
at <ScriptBlock>, /Users/runner/runners/2.166.4/work/1/s/test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1: line 909
909:             $res.CompletionMatches[0].CompletionText | Should -BeExactly "Test history completion"
```
Test-Connection.Ping.Force IPv4 with implicit PingOptions
```powershell
Expected exactly '10.79.9.126', but got 127.0.0.1.
at <ScriptBlock>, /Users/runner/runners/2.166.4/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 84
84:             $result[0].Address | Should -BeExactly $realAddress
```
Test-Connection.TraceRoute.TraceRoute works
```powershell
Expected exactly '10.79.9.126', but got 127.0.0.1.
at <ScriptBlock>, /Users/runner/runners/2.166.4/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 258
258:             $result[0].TargetAddress | Should -BeExactly $realAddress
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-windows`
TabCompletion.Miscellaneous completion tests.Test history completion
```powershell
Expected strings to be the same, but they were different.
Expected length: 23
Actual length:   640
Strings differ at index 0.
Expected: 'Test history completion'
But was:  '$env:POWER...'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 909
909:             $res.CompletionMatches[0].CompletionText | Should -BeExactly "Test history completion"
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-linux`
TabCompletion.Miscellaneous completion tests.Test history completion
```powershell
Expected strings to be the same, but they were different.
Expected length: 23
Actual length:   478
Strings differ at index 0.
Expected: 'Test history completion'
But was:  '$env:POWER...'
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1: line 909
909:             $res.CompletionMatches[0].CompletionText | Should -BeExactly "Test history completion"
```

PoshChan replied:
@vexx32, your last commit had 3 failures in `PowerShell-CI-macos`
TabCompletion.Miscellaneous completion tests.Test history completion
```powershell
Expected strings to be the same, but they were different.
Expected length: 23
Actual length:   598
Strings differ at index 0.
Expected: 'Test history completion'
But was:  '$env:POWER...'
at <ScriptBlock>, /Users/runner/runners/2.166.4/work/1/s/test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1: line 909
909:             $res.CompletionMatches[0].CompletionText | Should -BeExactly "Test history completion"
```
Test-Connection.Ping.Force IPv4 with implicit PingOptions
```powershell
Expected exactly '10.79.9.126', but got 127.0.0.1.
at <ScriptBlock>, /Users/runner/runners/2.166.4/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 84
84:             $result[0].Address | Should -BeExactly $realAddress
```
Test-Connection.TraceRoute.TraceRoute works
```powershell
Expected exactly '10.79.11.249', but got 127.0.0.1.
at <ScriptBlock>, /Users/runner/runners/2.166.4/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 258
258:             $result[0].TargetAddress | Should -BeExactly $realAddress
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-windows`
TabCompletion.Miscellaneous completion tests.Test history completion
```powershell
Expected strings to be the same, but they were different.
Expected length: 23
Actual length:   640
Strings differ at index 0.
Expected: 'Test history completion'
But was:  '$env:POWER...'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 909
909:             $res.CompletionMatches[0].CompletionText | Should -BeExactly "Test history completion"
```

iSazonov commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -2029,36 +2029,34 @@ private static object NullRule(CallSite site, object obj)
-            // If there are any fields, the type is mutable.
-            if (type.GetFields(BindingFlags.Public | BindingFlags.Instance).Any())
+            foreach (var attribute in type.GetCustomAttributes(inherit: false))
             {
-                return true;
+                if (attribute.GetType().FullName.Equals(
+                    "System.Runtime.CompilerServices.IsReadOnlyAttribute",
+                    StringComparison.Ordinal))
+                {
+                    return false;
```
The custom attribute targets old runtimes.
It is for edge scenarios to target old runtimes with modern compiler.
It will lose value over time.
So the addition could only have value for Windows PowerShell but not for PowerShell Core and we can remove it.

vexx32 replied:
Binary modules targeting `netstandard2.0` will still generate the custom attribute as the compiler can't be sure the attribute will be present when the DLL is loaded.

iSazonov replied:
>  Doing so effectively means we no longer support cross-compatibility with WinPS modules at all

No :-) In this case, we do not need to look back. What is current status of the code? 
- no security issues
- no performance issues
- no functional issues (do you know real broken scenarios? I think they don’t exist at all because it’s all for performance)

So only our intention is _to benefit from modern runtime to reduce false positives_.
I think "world wide plan" is to move community to .Net 6.0 LTS. That means Netstandard2.1. In any case, the old code will _decrease_ over time, and if we did not address (the performance) it earlier, then it does not make sense to address it to the future paying with performance due to new code without an urgent need.

vexx32 replied:
Not sure I follow you through to the conclusion there. Are you saying we should consciously break the compatibility with netstandard2.0 here and now, or are you saying we should leave behind the old, ineffective code as a fallback?

If you're saying it's not worth changing at all... I don't really agree. PS has to handle mutable from immutable structs a bit differently in a few cases that I see in the code. Being able to yield less false positives from this method should (in theory) allow the engine to take the more efficient code path more often -- though, certainly, the cases where we need to be working with structs in PS directly aren't that common.

iSazonov replied:
> Are you saying we should consciously break the compatibility with netstandard2.0 here and now, or are you saying we should leave behind the old, ineffective code as a fallback?

We can not break "compatibility with netstandard2.0" because the method is only for performance.
Yes, I ask you do not see back because it makes no sense - we have not real issues and real scenario to address and measure - measure it mandatory for performance changes. False positives on the old way will be reduced over time by themselves.

vexx32 replied:
That's not a bad point to make, I suppose. I don't really agree -- that method of checking is one of the fallback ones, so it's not going to be used in the primary case and thus is of secondary importance at best. I guess I don't see a compelling reason _not_ to check for that as well.

iSazonov replied:
>  I don't see a compelling reason

Do you already understand what you have done a breaking change and probably huge breaking change?
Interestingly, now that you have caught a bear, what will you do with it? :-)

SeeminglyScience replied:
@iSazonov 

> Do you already understand what you have done a breaking change and probably huge breaking change?
> Interestingly, now that you have caught a bear, what will you do with it? :-)

I don't see any breaking changes here.  Can you provide an example?

iSazonov replied:
CIs fail on this. To save your time - it is Datetime.

vexx32 replied:
What do you mean? `[datetime]` declares the correct attribute and will be registered as not mutable by this code correctly. Are you saying it's somehow mutable?

iSazonov replied:
Yes.

vexx32 replied:
You're gonna have to elaborate on that one and why it affects this, I'm afraid. I don't follow. Datetime isn't mutable as far as I'm aware.

iSazonov replied:
What does the method return for the type before and after the change?

vexx32 replied:
If the method was returning `true` in this case before this change, then the method is actually wrong. That's part of why I made this PR, because the assumptions this method was using are very flawed. We can pretty easily verify that datetime is immutable:

![image](https://user-images.githubusercontent.com/32407840/80826382-1c171080-8bb0-11ea-8928-a7730bc11cc1.png)

None of its fields can be modified; it's not mutable. Any methods on the type simply return a new instance, and don't modify the original. The fact that it's decorated with the appropriate compiler attribute also confirms this.

vexx32 replied:
If we have code that is relying on these false assumptions somehow, I'm happy to fix that too. I intend to investigate the CI failures tonight.

SeeminglyScience replied:
> CIs fail on this. To save your time - it is Datetime.

The only thing that is going to change with `datetime` is that when you assign one to a variable it will be done faster.

The current behavior is that it will be detected as mutable and a pointless copy will be done to ensure changes aren't persisted.  But it's immutable, there will be no changes, so no copy is required.

iSazonov replied:
@vexx32 Please revert the questionable change. The PR will be merged for non-LTS version and will affected most of users only after next PowerShell LTS - in the time this will be not actual I believe.

vexx32 commented:
@PoshChan please retry all

PoshChan commented:
@vexx32, successfully started retry of `PowerShell-CI-static-analysis, PowerShell-CI-Windows, PowerShell-CI-macOS, PowerShell-CI-Linux`

iSazonov commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -2029,36 +2029,34 @@ private static object NullRule(CallSite site, object obj)
-            for (int index = 0; index < properties.Length; index++)
+            // Fallback: check all fields (public + private) to verify whether they're all readonly.
+            // If any field is not readonly, the value type is mutable.
+            foreach (var field in type.GetFields(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic))
             {
-                var property = properties[index];
-                if (property.CanWrite)
+                if (!field.IsInitOnly)
```
What is a private field is not readonly but a property (used the field) is?
I think we should revert the code and enhance a code for methods below to take into account new readonly methods.

vexx32 replied:
The private field can still mutate, e.g., if the type provides a method that modifies the private field. Without checking all the fields, we cannot be sure the struct is immutable.

Also, even if we could check it more heuristically like you suggest, it's likely that doing more complex logic would be more expensive, since we'd need to make several different checks and verify a wider scope of things than just checking all the fields.

iSazonov replied:
Please re-read my _counterexample_:
> What is a private field is not readonly but a property (used the field) is?
```c#
using System;
struct S
{
    private int a;
    S(int p)
    {
        a = 1;
    }
    public int GetInt => a;
    public readonly int GetIntMethod()
    {
        // we could a = 2 if not readonly
        return a;
    }
}
```
You code return true but must false. [See](https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgDOwBOBXAxsAAgGVUBvVAyggB0zADcBDYAUwLADtDGBuCqogApOhagEp+lciiqyCjAgF4CCPjKoBfSQRgBmdlwIBxFsACShxQD55a2XoKYWjACYB7DhACeBwifNcALKmABZuLoIS6lLasgD0cQQA7my4btgQLvJKBHDsAGYEHG6ETq4e3rFUMADsttpaKBpAA) then compare with `private readonly int a;`.

Also the method is all for performance so we shouldn't address all rare cases. We should only reduce false positives in original code.

>  even if we could check it more heuristically like you suggest

I mean new C# 8.0 feature https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/struct#readonly-instance-members

vexx32 replied:
Right, but that's a limited case, and doesn't account for (as an example) a method being provided that mutates the private field. We don't have a way to ensure that a settable field is not being set, so it will reduce the false positives more effectively to simply check for settable fields.

I'm not sure I necessarily agree with your reasoning that it's only worth improving in certain ways. If we choose to compare lists of properties and methods as well as fields, in many cases we will end up needing to make _more_ reflection calls than only iterating the fields.

iSazonov replied:
>  but that's a limited case

Whole the method is limited case because it is only for value types and most of calls will fast return on first condition `type.IsPrimitive`.
_But it is BINDER code that is very-very sensitive_ and we need strong motivation to change it. Currently no use case exists we should address and no reliable tests and measures too.

vexx32 replied:
Right, that's why the current flawed assumptions are concerning.

But if it helps, I can add a case for primitives/enums back in, it most likely would short-circuit things a bit quicker.

vexx32 replied:
Added the short-circuiting for enums and primitives back in and tidied things up a little bit.

PoshChan commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -284,7 +284,7 @@ internal static BindingRestrictions CombineRestrictions(this DynamicMetaObject t
             var targetTemp = Expression.Variable(target.Expression.Type);
```
@vexx32, your last commit had 1 failures in `PowerShell-CI-linux`
TabCompletion.Miscellaneous completion tests.Test history completion
```powershell
Expected strings to be the same, but they were different.
Expected length: 23
Actual length:   478
Strings differ at index 0.
Expected: 'Test history completion'
But was:  '$env:POWER...'
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1: line 909
909:             $res.CompletionMatches[0].CompletionText | Should -BeExactly "Test history completion"
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-macos`
TabCompletion.Miscellaneous completion tests.Test history completion
```powershell
Expected strings to be the same, but they were different.
Expected length: 23
Actual length:   598
Strings differ at index 0.
Expected: 'Test history completion'
But was:  '$env:POWER...'
at <ScriptBlock>, /Users/runner/runners/2.166.4/work/1/s/test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1: line 909
909:             $res.CompletionMatches[0].CompletionText | Should -BeExactly "Test history completion"
```

PoshChan replied:
@vexx32, your last commit had 2 failures in `PowerShell-CI-windows`
Enter-PSHostProcess tests.By Process Id.Can enter using NamedPipeConnectionInfo
```powershell
Expected $true, but got $false.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Core\Enter-PSHostProcess.Tests.ps1: line 148
148:                     Should -BeTrue -Because (get-content $splat.FilePath -Raw)
```
TabCompletion.Miscellaneous completion tests.Test history completion
```powershell
Expected strings to be the same, but they were different.
Expected length: 23
Actual length:   633
Strings differ at index 0.
Expected: 'Test history completion'
But was:  '$env:POWER...'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 909
909:             $res.CompletionMatches[0].CompletionText | Should -BeExactly "Test history completion"
```

vexx32 committed:
:wrench: Improve detection of mutable value types

Currently, we have some flawed assumptions around what defines a mutable
value type. Rather than make a more complicated series of checks and
making assumptions, we can simply look at whether the type is defined
with the IsReadOnlyAttribute.

As a fallback, it's much safer and less complex to enumerate public and
private fields on the type, and verify whether they're all read-only.

Resolves #12430.

PoshChan commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -2029,36 +2029,38 @@ private static object NullRule(CallSite site, object obj)
 
```
@vexx32, your last commit had 1 failures in `PowerShell-CI-linux`
TabCompletion.Miscellaneous completion tests.Test history completion
```powershell
Expected strings to be the same, but they were different.
Expected length: 23
Actual length:   478
Strings differ at index 0.
Expected: 'Test history completion'
But was:  '$env:POWER...'
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1: line 909
909:             $res.CompletionMatches[0].CompletionText | Should -BeExactly "Test history completion"
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-macos`
TabCompletion.Miscellaneous completion tests.Test history completion
```powershell
Expected strings to be the same, but they were different.
Expected length: 23
Actual length:   598
Strings differ at index 0.
Expected: 'Test history completion'
But was:  '$env:POWER...'
at <ScriptBlock>, /Users/runner/runners/2.166.4/work/1/s/test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1: line 909
909:             $res.CompletionMatches[0].CompletionText | Should -BeExactly "Test history completion"
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-windows`
TabCompletion.Miscellaneous completion tests.Test history completion
```powershell
Expected strings to be the same, but they were different.
Expected length: 23
Actual length:   633
Strings differ at index 0.
Expected: 'Test history completion'
But was:  '$env:POWER...'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 909
909:             $res.CompletionMatches[0].CompletionText | Should -BeExactly "Test history completion"
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-linux`
TabCompletion.Miscellaneous completion tests.Test history completion
```powershell
Expected strings to be the same, but they were different.
Expected length: 23
Actual length:   478
Strings differ at index 0.
Expected: 'Test history completion'
But was:  '$env:POWER...'
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1: line 909
909:             $res.CompletionMatches[0].CompletionText | Should -BeExactly "Test history completion"
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-macos`
TabCompletion.Miscellaneous completion tests.Test history completion
```powershell
Expected strings to be the same, but they were different.
Expected length: 23
Actual length:   598
Strings differ at index 0.
Expected: 'Test history completion'
But was:  '$env:POWER...'
at <ScriptBlock>, /Users/runner/runners/2.166.4/work/1/s/test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1: line 909
909:             $res.CompletionMatches[0].CompletionText | Should -BeExactly "Test history completion"
```

PoshChan replied:
@vexx32, your last commit had 1 failures in `PowerShell-CI-windows`
TabCompletion.Miscellaneous completion tests.Test history completion
```powershell
Expected strings to be the same, but they were different.
Expected length: 23
Actual length:   633
Strings differ at index 0.
Expected: 'Test history completion'
But was:  '$env:POWER...'
at <ScriptBlock>, D:\a\1\s\test\powershell\Host\TabCompletion\TabCompletion.Tests.ps1: line 909
909:             $res.CompletionMatches[0].CompletionText | Should -BeExactly "Test history completion"
```

vexx32 commented:
@PoshChan please retry all

PoshChan commented:
@vexx32, did not find any matching pull request checks

iSazonov commented:
@vexx32 Do you want to continue?

vexx32 commented:
I'm unsure. Can we tag this for committee review, please?

Let me explain as best I can. @SeeminglyScience may be able to explain the bits I'm not especially clear on, I'm painfully aware I don't have a thorough understanding of everything that's going on.

1. **Currently**, our detection of mutable value types is simply broken. There's no other conclusion in my mind, it's literally completely broken;
2. Any but the very simplest and most basic structs will register as mutable and end up being copied unnecessarily, so we're wasting a decent amount of cycles copying value types we simply don't need to.

This is complicated by the fact that one of the methods that are using this mutable struct detection method treats them more differently than it should. In short:

1. If it reads the value type as mutable, it _unwraps PSObjects_ as part of the copy operation before passing it along to whatever's using it, and it is never rewrapped before being handed back.
    - I don't understand the full implications of this, but as I understand it this will cause performance degradation since there are many times we have to then look up the PSObject wrapper for that struct value/instance from the resurrection tables.
2. It doesn't do that for immutable structs, so these end up remaining wrapped in PSObject.

So, in other words... whether a struct is considered mutable or not (and we have a _lot_ of false positives) determines whether we waste potentially quite a lot of cycles looking up PSObject wrappers in the resurrection table for it _repeatedly_.

I'm of the opinion this should just be fixed; both code paths should be returning values the same way. Not sure if that should be wrapped or not, but both should be the same. Currently I'd say that ensuring all structs returned should be returned wrapped as that'll result in better performance overall.

_But_ since most structs currently are (incorrectly) registered as mutable, they end up being returned unwrapped. So some compiled cmdlets may notice the change -- actually, that's what's causing the CI failures. It's expecting a raw DateTime value when it should be checking for PSObject wrappers and grabbing the base object out of it to get the DateTime value. I don't know how long it's worked like this.

Finally, It's a little bit of an edge case for cmdlets to ever notice the difference; in order to see this difference, you have to be taking as input some kind of object that has `DateTime` as a property. If you take in `DateTime` as a parameter value, you'll never see this difference, it will always be unwrapped by the time you see it. The cmdlet in the repo that is registering this change accepts a PSObject as input and expects a `DateTime` property on it; handling this the way that seems sensible currently breaks that as it ends up with a `PSObject` wrapped around it in that instance.

I'm also going to tag @daxian-dbw for review as well. In my opinion, this is reasonably safe to change and save the performance cycles, but I'm also aware that it is a potentially noticeable change to the way this works, and it may break some compiled cmdlets that expect (for example) PSCustomObject / PSObject as input data, with one of the values being a struct that currently is incorrectly considered mutable.

I'm aware that's a long and potentially unclear explanation, please ask if you have any questions, I'll attempt to clarify as best I can. 💖

SeeminglyScience commented:
@vexx32 did a great job at explaining but let me try to summarize with an example

```powershell
# Current behavior that incorrectly detects DateTime as mutable
(Get-Date) -is [psobject]
# True
$a = Get-Date
$a -is [psobject]
# False

# Fixed behavior
(Get-Date) -is [psobject]
# True
$a = Get-Date
$a -is [psobject]
# True
```

The way that copy semantics are implemented, the `PSObject` is unwrapped, the struct is unboxed and boxed again, but **never rewrapped**.  In PowerShell this is meaningless, but a PowerShell aware (or focused) C# API may expect that a mutable struct (or rather, any non-primitive value type because detection is busted) will never be wrapped in a `PSObject`.  Personally I think such an API should *always* account for `PSObject`, but it's easy to imagine this behavior being accidently depended on.  Especially since PowerShell has at least one example of this.

Technically this behavior is undocumented and arguably not part of the breaking change contract, but it's also quite difficult to know exactly what the impact would be.  I think it's worth the change.  I'd also like to bring up the possibility of adding a rewrap step in the copy semantics if the mutable struct was originally wrapped in a `PSObject` (though maybe better suited for a new issue).

iSazonov commented:
We put several dangerous things in one PR and this PR will have unpredictable consequences.
I would explicitly separate them for discussion and conclusion:
- make Datetime immutable
- improve copy semantics for rewrapping in PSObject
- improve mutable struct detection based on latest C# capabilities
   - with/without taking into account old code
- ?

After that we could ask PowerShell Committee.

vexx32 commented:
> make Datetime immutable

What? It _is_ immutable. There's nothing we need to do here. If we make an explicit exception for DateTime in our handling code we're simply monkey-patching the obvious case in a way that will have about the same impact for that type anyway.

We can't affect this without also needing to address the other issues. I'm all for making the _changes_ piecemeal if we can, but I don't think a meaningful decision can be had unless we have a plan to act on already in place.

iSazonov commented:
> What? It is immutable. 

PowerShell detects it as mutable.

vexx32 commented:
Yes, and it's incorrect. We've covered this.

SeeminglyScience commented:
Just to clarify, this is the only issue that needs ruling on currently:

If we fix detection of immutability, structs that were previously (falsely) detected as mutable may in some cases be wrapped in a `PSObject` where in others they were simply a boxed struct.  This could break some scenarios where a C# API traverses a psobject with a property containing one of these structs.

For example, a API like this:

```csharp
bool HasModifiedTime(PSObject pso)
{
     return pso.Properties["ModifiedTime"]?.Value is DateTime;
}
```

Should instead look like:

```csharp
bool HasModifiedTime(PSObject pso)
{
    var value = pso.Properties["ModifiedTime"]?.Value;
    return value is PSObject inner
        ? inner.BaseObject is DateTime
        : value is DateTime;
}
```

Or really most cases will look something like (which will still work):

```csharp
bool GetModifiedTime(PSObject pso)
{
    return LanguagePrimitives.ConvertTo<DateTime>(pso.Properties["ModifiedTime"]?.Value);
}
```

Basically, is this change acceptable given that most code will (or should) look like the second or third example? All other points can be discussed in separate issues.

vexx32 commented:
I feel I've made my questions pretty clear... I think this either needs a committee review or perhaps @rjmholt and/or @daxian-dbw to take a close look and examine what best to do here (or whether we're better off leaving it as is).

rjmholt commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -2029,36 +2029,38 @@ private static object NullRule(CallSite site, object obj)
+            // If the builtin attribute is not present, check for a custom attribute from by the compiler. If the
+            // library targets netstandard2.0, the compiler can't be sure the attribute will be provided by the runtime,
+            // and defines its own attribute of the same name during compilation.
+            foreach (var attribute in type.GetCustomAttributes(inherit: false))
```
`var` -> type

vexx32 replied:
🤔 `GetCustomAttributes()` returns only `object[]`. Is it safe to type this var as `Attribute`?

rjmholt replied:
If `object` is all it guarantees, then I think `object` makes sense

rjmholt commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -2029,36 +2029,38 @@ private static object NullRule(CallSite site, object obj)
+            foreach (var attribute in type.GetCustomAttributes(inherit: false))
             {
-                return true;
+                if (attribute.GetType().FullName.Equals(
+                    "System.Runtime.CompilerServices.IsReadOnlyAttribute",
+                    StringComparison.Ordinal))
```
What's the reason for needing to do a type name comparison rather than a simple type comparison here?

SeeminglyScience replied:
> What's the reason for needing to do a type name comparison rather than a simple type comparison here?

@rjmholt if you target `netstandard2.0` and use `readonly struct`s then Roslyn embeds the attribute definition in the assembly, so type identity will only be reliable for assemblies targeting netcoreapp3.0+ (approximately, don't remember when that was introduced).

rjmholt replied:
Ok got it -- might be worth a comment so that someone doesn't "optimise" it

vexx32 replied:
I already added comments for this just above (2040-2043) - should I move those closer to this line?

rjmholt replied:
Ah, I would just add something like "so we must check for the type by name rather than by reference" after the bit about the compiler

vexx32 commented on src/System.Management.Automation/engine/runtime/Binding/Binders.cs:
```diff
@@ -2029,36 +2029,38 @@ private static object NullRule(CallSite site, object obj)
+            // If the builtin attribute is not present, check for a custom attribute from by the compiler. If the
+            // library targets netstandard2.0, the compiler can't be sure the attribute will be provided by the runtime,
+            // and defines its own attribute of the same name during compilation.
+            foreach (var attribute in type.GetCustomAttributes(inherit: false))
```
```suggestion
            foreach (object attribute in type.GetCustomAttributes(inherit: false))
```

vexx32 committed:
:recycle: Address Rob's feedback

Use `object` instead of `var`

:memo: clarify comment on attribute check

vexx32 committed:
:bug: Resolve issue in history input handling

This resolves the fragile conversions in the history input handling.

Essentially, anything taking input in this manner
(values nested in a hashtable / other loosely-typed property bag)
need to be careful to call conversion APIs or always be careful to check
for and retrieve the BaseObject if the input happens to be PSObject.

This was previously "fine" largely by accident, and would only have
been noticed previously for unrealistically simple structs which were
mistakenly not rewrapped as PSObject.

I additionally reframed the other conversions in this method to follow
a common pattern and use existing functionality in LanguagePrimitives.

rjmholt commented:
Merging this in so we can get early validation for it in the next preview

iSazonov commented:
> Merging this in so we can get early validation for it in the next preview

Have you a plan for the validation? I still have concerns about the implementation.

rjmholt commented:
@iSazonov can you summarise your concerns here? With the changes in a release we might be able to get more users trying this out. We can then set up a decision on whether to revert this, require changes or leave it as is before the next release.

iSazonov commented:
> @iSazonov can you summarize your concerns here for other readers?

Oh, every step in this change raises questions. You can see it in my comments above.
- we did not measure performance on the hot path, we can get a regression.
- we changed the current behavior without a conclusion
- we added new behavior without impact analysis.

Ideally, we should have seen good perf win in some scenarios. What is the scenarios and where is perf tests? Do we cover all new .Net features?
